### PR TITLE
[tvm][any] broadcast with values other than one

### PR DESCRIPTION
**Pull request by zhiics**
This PR makes relay Any type broadcast with values other than one, i.e.:

```python
x = relay.var('x', shape=(3, 2), dtype="float32")
y = relay.var('y', shape=(relay.Any(), 2), dtype="float32")
relay.add(x, y)
```

It makes all symbolic vars as `auto_broadcast` if they haven't been bound to compact buffer before.

@icemelon9 @yzhliu @tqchen @jroesch

### CODE REVIEW

**Comment by jroesch** [comment]
I am a bit confused by the semantics, how does this work? for example imagine that I have two any dimensions which will be (say 20, and 15 at runtime) it doesn't make sense to broad cast against non-1 dimensions.

**Comment by zhiics** [comment]
@jroesch sorry. I think the example I gave missed something. I updated it.

**Comment by zhiics** [comment]
close for now. need more consideration about the auto_broadcast binding.

**Comment by junrushao1994** [comment]
Hey this is super helpful!

Btw, I notice that in the compile engine it assumes that checked type exists. Does it mean that for now we are only able to deal with the case that dims are fixed and certain dimensions are Any?

**Comment by zhiics** [comment]
@junrushao1994 Yes, you are right, only the bound is dynamic but rank is fixed.

**Comment by kevinthesun** [comment]
It will generate runtime error when actual shape is incompatible?

**Comment by zhiics** [comment]
@kevinthesun Yes, both the interpreter and VM would fail if the real values are not incompatible. I can probably add a few tests for them as well.

**Comment by icemelon9** [comment]
> It will generate runtime error when actual shape is incompatible?

Yes, the shape function will perform the type checking at runtime (see https://github.com/dmlc/tvm/blob/master/python/tvm/relay/op/_tensor.py#L130)

**Comment by zhiics** [comment]
@icemelon9 yeah, that might be a different between Any and tvm symbolic var.

File: python/tvm/build_module.py [hunk 1]
```diff
@@ -290,12 +293,15 @@ def get_binds(args, binds=None):
     arg_list = []
     for x in args:
         if isinstance(x, tensor.Tensor):
+            any_dim = any(isinstance(i, expr.Var) for i in x.shape)
+            buffer_type = "auto_broadcast" if any_dim and not compact else ""
```
Comment (kevinthesun) [inline python/tvm/build_module.py]: Just curious, what does it mean when any_dim and compact are both true?

Comment (zhiics) [reply python/tvm/build_module.py]: `any_dim` indicates if it is symbolic. Here, we don't really distinguish relay Any and tvm Var. `compact` means that a buffer has already declared as compact type before (i.e. with attribute attr::buffer_bind_scope) and bound to the variable `x` using `AttrStmt`. If so, we shouldn't not declare a strided buffer and bind it to `x`.

**Comment by zhiics** [comment]
@yzhliu Yeah, I actually thought about replacing it as well. But I ended up skipping it because it looks there is not much difference. I think we probably want to keep compact memory as much as possible because auto broadcast will bring in the if_then_else guard which may not as performant as the compact version. How do you think?

**Comment by zhiics** [comment]
@icemelon9 @yzhliu @kevinthesun I think we've probably forgotten this PR.

**Comment by junrushao1994** [comment]
Is this ready for review?

**Comment by zhiics** [comment]
@junrushao1994 yes

**Comment by yzhliu** [comment]
> @yzhliu Yeah, I actually thought about replacing it as well. But I ended up skipping it because it looks there is not much difference. I think we probably want to keep compact memory as much as possible because auto broadcast will bring in the if_then_else guard which may not as performant as the compact version. How do you think?

so it won't work for those ops written in hybrid script, right?

**Comment by zhiics** [comment]
@yzhliu Thanks. Removed the binding in hybrid script as it will go through `lower` anyway.

**Pull request merged**